test/test_nn.py

```diff
@@ -2582,6 +2582,127 @@ def set_extra_state(self, state):
             self.assertEqual(m.state_dict(), m2.state_dict())
             self.assertEqual(m.foo, m2.foo)
 
+    def test_load_state_dict_assign_meta(self):
+        class MyModule(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.fc1 = nn.Linear(3, 5)
+                self.bn = nn.BatchNorm1d(5)
+
+            def forward(self, input):
+                return self.bn(self.fc1(input))
+
+        net = MyModule()
+        state_dict = net.state_dict()
+
+        with torch.device('meta'):
+            net_meta = MyModule()
+
+        with self.assertWarnsRegex(
+            expected_warning=UserWarning,
+            expected_regex="If using ``load_state_dict`` with ``assign=True``, "
+            "please make sure to initialize the optimizer or",
+        ):
+            net_meta.load_state_dict(state_dict, assign=True)
+
+        # Make sure parameters and persistent buffers were assigned
+        net_meta_state_dict = net_meta.state_dict()
+        for key in state_dict.keys():
+            self.assertTrue(state_dict[key].data_ptr() == net_meta_state_dict[key].data_ptr())
```

Comment:
Can't this be `state_dict[key] is net_meta_state_dict[key]` if you pass `keep_vars=True` when getting the state_dicts?

Reply:
Hm, I guess because [here](https://github.com/pytorch/pytorch/blob/e6e9fd9e7eeda6668b26557f2ad07d4751af68ca/torch/nn/modules/module.py#L1968) during assignment I always wrap in `torch.nn.Parameter` (which could take either an `nn.Parameter` or a Tensor as input), they are actually different python objects

Reply:
Ho that is interesting. Commented below but I think it would be a great property for us to preserve during assignment.
Making new Tensor object is cheap and we do it easily. But when subclasses are involved. We don't know how expensive a new object will be.